docs/specs/code/LOCK_WALLET_WITH_PIN.md

```diff
@@ -0,0 +1,59 @@
+# Abstract
+
+Allows user to lock their wallet with a pin like someone can do in iOS / Android.
+
+# Motivation
+
+Yoroi is completely secure (impossible to do tx without the password), but we want to add a lock screen similar to what phones have so we can protect the user privacy (balance and transactions) in case that someone else has access to the computer. This has a been a constant request from multiple people.
+
+# Proposal
+
+## How it should work
+
+The user should have like a possibility to click an icon in the topbar (a lock) and then the app will be locked; in the locked state, the user can’t see any information except for a dialog where a PIN is asked (we will name this pin/ pincode to avoid confusion with the spending password)
+
+## Requirements
+
+- add a dummy icon in the top right part of the topbar -> activate a locking screen

```

Comment:
So even when PIN code is activated in settings - user has to explicitly click a button to activate the lock? (@nicarq)

I would suggest making an additional option, like "Lock automatically when closed". Which might be disabled by default, but when enabled - Yoroi is automatically locked on every exit. 

Technically it might be achievable if we keep local storage always encrypted and when user "logs in" - we keep the decryption key in memory, throughout the working of the app, using it at every storage request, or something like that. And when app is closed - key is automatically cleared from memory.

Reply:
Tbh, I'm not sure how storage encryption would work at all in case when Yoroi does NOT auto-lock on exit. Does it mean we would generally keep storage decrypted, but then encrypt it when user clicks the lock button? And then decrypt it on "log in" and keep decrypted until next "lock"?

Reply:
Good point. I was thinking it would just be a toggle on/off but you're right that some users may want auto-lock either when they close to the tab or focus a different tab. We could implement this with a hook on tab close, but it wouldn't work if the computer suddenly turns off or the browser suddenly closes. You could do a best-effort work around by locking the app right away if it detects the browser was closed without locking.
You could do this by making that every time you open Yoroi it saves a GUID to localcache (which I think we should do anyway) and save the session GUID to a separate field `lastUnlockGuid` in localcache every time a user unlocks. If the GUID in localcache doesn't match the current GUID the screen locks.